lib/cdo/buffer.rb

```diff
@@ -0,0 +1,155 @@
+    # Implement in subclass.
+    # @param [Array<Object>] objects
+    def flush(objects)
+    end
+
+    # Calculate the total size of a batch of objects.
+    # Override in subclass when using 'batch_size' limit.
+    # @param [Array<Object>] objects
+    # @return [Numeric] size of objects
+    def size(objects)
+      1
+    end
+
+    # Add an object to the buffer.
+    # @raise [ArgumentError] when the object exceeds batch size
+    # @param [Object] object
+    def buffer(object)
+      reset_if_forked
+      if (size = size([object])) > @batch_size
+        raise ArgumentError, "Object size (#{size}) exceeds batch size (#{@batch_size})"
+      end
+      @buffer << BufferObject.new(object, now)
+      schedule_flush
+    end
+
+    # Flush existing buffered objects.
+    # @param [Float] timeout seconds to wait for buffered objects to finish flushing.
+    def flush!(timeout = Float::INFINITY)
+      reset_if_forked
+      timeout_at = now + timeout
+      until (wait = timeout_at - now) < 0 || @buffer.empty?
+        @log.info "Flushing #{self.class}, waiting #{wait} seconds"
+        schedule_flush(true)
+        # Block until the pending flush is completed or timeout is reached.
+        @scheduled_flush.wait(wait.infinite? ? nil : wait)
+      end
+    end
+
+    private
+
+    # Track time each object was added to the buffer.
+    BufferObject = Struct.new(:object, :added_at)
+
+    def now
+      Concurrent.monotonic_time
+    end
+
+    # Schedule a flush in the future when the next batch is ready.
+    # @param [Boolean] force flush batch even if not full.
+    def schedule_flush(force = false)
+      delay = batch_ready(force)
+      if @scheduled_flush.pending?
+        @scheduled_flush.reschedule(delay)
+      else
+        @scheduled_flush = Concurrent::ScheduledTask.execute(delay) do
+          flush_batch
+          schedule_flush unless @buffer.empty?
+        end
+      end
+    end
+
+    # Determine when the next batch of existing buffered objects will be ready to be flushed.
+    # @param [Boolean] force flush batch even if not full.
+    # @return [Float] Seconds until the next batch can be flushed.
+    def batch_ready(force)
+      return Float::INFINITY if @buffer.empty?
+
+      # Wait until max_interval has passed since the earliest object to flush a non-full batch.
+      earliest = @buffer.first.added_at
+      wait = @max_interval - (now - earliest)
+
+      # Flush now if the batch is full or when force flushing.
+      wait = 0.0 if force ||
+        size(@buffer.map(&:object)) >= @batch_size ||
+        @buffer.length >= @batch_count
+
+      # Wait until min_interval has passed since the last flush.
+      min_delay = @min_interval - (now - @last_flush.to_f)
+
+      [0.0, wait, min_delay].max.to_f
+    end
+
+    # Flush a batch of objects from the buffer.
+    def flush_batch
+      @last_flush = now
+      flush(take_batch.map(&:object))
+    rescue => e
+      Honeybadger.notify(e)
+      raise
```

Comment:
What happens when we `raise` this exception? Are we planning on having something catching it?

Reply:
I think I had the initial idea that this could raise an exception when `flush!` was called explicitly to synchronously flush buffers, but I think it's simpler to just have any exceptions just be asynchronously raised in honeybadger for consistency so I removed the re-`raise` line.